Modularize workflows by creating reusable GitHub Actions

mamu0 opened the pull request:
This PR introduces some new **GitHub Actions** and modifies existing workflows to use them. The main goal was to reduce code duplication across workflows by extracting common functionality into reusable actions.

#### Actions created:
- `azure-login`: For standardized authentication with Azure
- `terraform-setup`: For Terraform version configuration
- `upload-artifact`: For uploading artifacts in pipelines
- `download-artifact`: For downloading artifacts in pipelines
- `node-setup`: For configuring Node.js environment
- `get-all-modules` and `get-modified-modules`: For managing Terraform modules
- `slack-notification`: For sending on slack channel a notification

#### Other changes:
- Improved `get-terraform-version` action with new terraform version input
- renamed `cdn-code-deploy` to maintein convention
- deleted `get_terraform_version.yaml`, `get_all_modules.yaml` and `get_modified_modules.yaml` workflow to use new actions

### Benefits:
- Easier maintenance: changes in a single point instead of multiple workflows (E.g. the commit `sha`)
- Consistent behavior across different workflows
- Cleaner and more readable code
- Easier to add new workflows that reuse the same functionality

Resolves: CES-737

mamu0 commented on .github/actions/azure-login/action.yaml:
```diff
@@ -0,0 +1,13 @@
+runs:
+  using: "composite"
+
+  steps:
+    - name: Azure Login
+      uses: azure/login@a65d910e8af852a8061c627c456678983e180302 # v2
+      with:
+        client-id: ${{ env.ARM_CLIENT_ID }}
+        tenant-id: ${{ env.ARM_TENANT_ID }}
+        subscription-id: ${{ env.ARM_SUBSCRIPTION_ID }}
```
As I explained in the description, one of the advantages of using actions is having a single point of modification. For example, in this case, to avoid having different SHAs across workflows—resulting in inconsistencies—it is sufficient to update it here to apply the change everywhere

lucacavallaro replied:
IMO we can solve this problem in a different way: we should se up a procedure that updates the actions to their latest available version. We don't need to manage multiple version, the only valid version is the latest available.

mamu0 replied:
I’m not sure if we have the autonomy to decide to stop using SHAs and switch to versioning instead (@gunzip?). 
But, for consistency, it is still advisable to specify a `version` rather than using the `latest`, so having a single point for updates remain more convenient.

In any case, I noticed that the step was usually defined like this:
```
- name: Azure Login
  uses: azure/login@a65d910e8af852a8061c627c456678983e180302 # v2
  with:
    client-id: ${{ env.ARM_CLIENT_ID }}
    tenant-id: ${{ env.ARM_TENANT_ID }}
    subscription-id: ${{ env.ARM_SUBSCRIPTION_ID }}
  env:
    ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
    ARM_TENANT_ID: ${{ secrets.ARM_TENANT_ID }}
    ARM_SUBSCRIPTION_ID: ${{ secrets.ARM_SUBSCRIPTION_ID }}
```
Using the action makes it simpler since only the environment variables are needed:
```
- uses: ./.github/actions/azure-login
  name: Azure Login
  env:
    ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
    ARM_TENANT_ID: ${{ secrets.ARM_TENANT_ID }}
    ARM_SUBSCRIPTION_ID: ${{ secrets.ARM_SUBSCRIPTION_ID }}
```
Additionally, if these `envs` are defined at the _job_ or `workflow` level, there’s no need to specify them again in the step, further simplifying the configuration.

lucacavallaro replied:
@mamu0 I meant to use the latest SHA available, don't switch to semver

mamu0 replied:
@pagopa/engineering-team-devex How do we want to proceed with this? Should we keep the steps as they were and maybe evaluate a tool (like the one suggested by @gunzip), or should we use actions?

mamu0 commented on .github/actions/download-artifact/action.yaml:
```diff
@@ -0,0 +1,26 @@
+runs:
+  using: "composite"
+
+  steps:
+      - name: Download Artifact
+        uses: actions/download-artifact@6b208ae046db98c579e8a3aa621ab581ff575935 # v4.1.1
+        with:
+          name: ${{ inputs.bundle_name }}
+          path: ${{ inputs.file_path }}
```
Same concept as the previous response, so we have a single update point since it is used multiple times

mamu0 commented on .github/actions/get-terraform-version/action.yaml:
```diff
@@ -1,14 +1,21 @@
 description: Retrieve the Terraform version from the `.terraform-version` file or return a default version.
+
 inputs:
+  terraform_version:
+    description: "Specify the Terraform version to use, if not set, it will be read from .terraform-version file"
+    type: string
+    required: false
   default_version:
     description: "Default Terraform version to use if the `.terraform-version` file is not found."
     required: false
     default: "1.10.4"
```
As it was previously configured, regardless of the `default_version` value indicated, priority was always given to the `.terraform_version` file. The input `terraform_version` allows for an override and the ability to specify a different version.

The usefulness of this feature is practical since, being an Action rather than a Workflow, the more configurations there are, the more it can be reused.

A use case could also be a simple test, for example, running two workflows simultaneously in the same repository to observe behavior based on different Terraform versions.

mamu0 commented on .github/actions/node-setup/action.yaml:
```diff
@@ -0,0 +1,41 @@
+        restore-keys: |
+          ${{ runner.os }}-turbo-
+
+    - name: Install dependencies
+      shell: bash
+      run: yarn install --immutable
+      working-directory: ${{ inputs.working_dir }}
```
Are you sure? I included it because the command is exactly the same across all workflows. Regardless of the different results it may produce in certain cases, since it is always written the same way and consistently used, it makes sense to keep it as part of the setup action, avoiding the need to repeat the same step multiple times

gunzip replied:
I'm wondering if we need to differentiate between production and development installs (ie. avoiding dev dependencies)

gunzip replied:
if we can set the `NODE_ENV` environment in the caller and make it reach this point it's ok as is, see 

https://classic.yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-production-true-false

lucacavallaro replied:
@gunzip there is no way esclude devDependencies right now - yarn 4 support this mode only using workspace focus that doesn't work well with turbo prune

I think is out of scope right now

mamu0 commented on .github/actions/slack-notification/action.yaml:
```diff
@@ -0,0 +1,77 @@
+          title="${{ inputs.title }}"
+        fi
+
+        if [ "${{ inputs.slack_webhook_url }}" == "" ]; then
+          if [ "${{ env.SLACK_WEBHOOK_URL }}" == "" ]; then
+            echo "No Slack Webhook URL provided. Exiting."
+            exit 1
+          fi
+          slack_webhook_url="${{ env.SLACK_WEBHOOK_URL }}"
+        else
+          slack_webhook_url="${{ inputs.slack_webhook_url }}"
+        fi
```
don’t have a strong opinion on this. I just thought that this way, anyone using the action could choose between the classic configuration (the environment) or a more direct approach (the input). However, if all of you prefer a single method, I can update it to use only the input

gunzip replied:
agree

gunzip reviewed (commented):
**question**: are the deleted files referenced somewhere?

mamu0 commented:
> **question**: are the deleted files referenced somewhere?

no, two were used improperly, as they were two actions made as workflow, so I converted and replaced them, one was unused and had already been converted back into an action

kin0992 commented on .github/workflows/function_app_deploy.yaml:
```diff
@@ -141,16 +139,12 @@ jobs:
-          subscription-id: ${{ env.ARM_SUBSCRIPTION_ID }}
+          bundle_name: ${{ env.BUNDLE_NAME }}
+          
+      - uses: ./.github/actions/azure-login
+        name: Azure Login
```
**nitpick**: just to be consistent

```suggestion
      - name: Azure Login
         uses: ./.github/actions/azure-login
```
I'd keep the same order everywhere

Krusty93 commented on .github/actions/terraform-setup/action.yaml:
```diff
@@ -0,0 +1,39 @@
+      with:
+        terraform_version: ${{ steps.get-version.outputs.terraform_version }}
+
+    - name: Terraform Init
```
**nitpick**: Not sure to include `terraform init` here. It is a simple call to make and decrease flexibility in the case where caller would like to do some action before initialising Terraform

mamu0 replied:
I see what you mean. In your opinion, what might need to be done between the setup and init? I’m asking because before removing the step, it could be useful to evaluate whether adding other configurable optional steps would be beneficial or if keeping init separate would actually be more convenient

Krusty93 commented on .github/workflows/infra_apply.yaml:
```diff
@@ -100,29 +100,14 @@ jobs:
+        name: Azure Login
 
-      - name: Set Terraform Version
+      - name: Terraform Setup and Init
```
In reference to my [previous comment](https://github.com/pagopa/dx/pull/356/files#r1998809670), this name is showing the misunderstanding that may arise from the user

gunzip commented on .github/actions/get-terraform-version/action.yaml:
```diff
@@ -1,14 +1,21 @@
 description: Retrieve the Terraform version from the `.terraform-version` file or return a default version.
+
 inputs:
+  terraform_version:
```
ok
